### PR TITLE
Added support of WB32 DFU for MacOs.

### PR DESCRIPTION
**Pull request by wb-Joy**
Added support of WB32 DFU for MacOs.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

*

### CODE REVIEW

**Comment by fauxpark** [comment]
Package should be submitted to Homebrew and binary taken from there.
https://github.com/homebrew/homebrew-core

See #292

File: macos/QMK Toolbox/USB/Bootloader/BootloaderType.h [hunk 1]
```diff
@@ -11,6 +11,7 @@ typedef enum BootloaderType : NSUInteger {
     BootloaderTypeQMKDFU,
     BootloaderTypeSTM32DFU,
     BootloaderTypeSTM32Duino,
+    BootloaderTypeWB32DFU,
     BootloaderTypeUSBAsp,
     BootloaderTypeUSBTinyISP,
```
Comment (fauxpark) [inline macos/QMK Toolbox/USB/Bootloader/BootloaderType.h:16]:
```suggestion
    BootloaderTypeUSBAsp,
    BootloaderTypeUSBTinyISP,
    BootloaderTypeWB32DFU,
```

File: macos/QMK Toolbox/USB/USBListener.m [hunk 2]
```diff
@@ -14,6 +14,7 @@
 #import "LUFAMSDevice.h"
 #import "STM32DFUDevice.h"
 #import "STM32DuinoDevice.h"
+#import "WB32DFUDevice.h"
 #import "USBAspDevice.h"
 #import "USBTinyISPDevice.h"
```
Comment (fauxpark) [inline macos/QMK Toolbox/USB/USBListener.m:19]:
```suggestion
#import "USBAspDevice.h"
#import "USBTinyISPDevice.h"
#import "WB32DFUDevice.h"
```

File: macos/QMK Toolbox/USB/USBListener.m [hunk 3]
```diff
@@ -150,6 +151,8 @@ - (void)stop {
             return [[STM32DFUDevice alloc] initWithUSBDevice:usbDevice];
         case BootloaderTypeSTM32Duino:
             return [[STM32DuinoDevice alloc] initWithUSBDevice:usbDevice];
+        case BootloaderTypeWB32DFU:
```
Comment (fauxpark) [inline macos/QMK Toolbox/USB/USBListener.m:154]: Move below `BootloaderTypeUSBTinyISP`

File: macos/QMK Toolbox/USB/USBListener.m [hunk 4]
```diff
@@ -187,6 +190,11 @@ - (BootloaderType)deviceTypeForVendorID:(ushort)vendorID productID:(ushort)produ
                 return BootloaderTypeSTM32DFU;
             }
             break;
+        case 0x342D: // WestBerryTech
```
Comment (fauxpark) [inline macos/QMK Toolbox/USB/USBListener.m:193]: Move below `0x314B`

File: macos/QMK Toolbox/USB/Bootloader/WB32DFUDevice.m [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    return self;
+}
+
+-(void)flashWithMCU:(NSString *)mcu file:(NSString *)file {
+    if([[[file pathExtension] lowercaseString] isEqualToString:@"bin"]) {
+        [self runProcess:@"wb32-dfu-updater_cli" withArgs:@[@"--toolbox-mode", @"--dfuse-address", @"0x08000000", @"--download", file]];
+    } else if([[[file pathExtension] lowercaseString] isEqualToString:@"hex"]) {
+        [self runProcess:@"wb32-dfu-updater_cli" withArgs:@[@"--toolbox-mode", @"--download", file]];
+    } else {
+        [self printMessage:@"This file format is not the type supported by wB32-dFU-updater_cli!" withType:MessageType_Error];
```
Comment (fauxpark) [inline macos/QMK Toolbox/USB/Bootloader/WB32DFUDevice.m:20]:
```suggestion
        [self printMessage:@"Only firmware files in .bin or .hex format can be flashed with wb32-dfu-updater_cli!" withType:MessageType_Error];
```

File: macos/QMK Toolbox.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -366,6 +374,7 @@
 				09522BBB1F6216BA00AEBC5E /* avrdude.conf in Resources */,
 				09522BB31F61E32700AEBC5E /* mcu-list.txt in Resources */,
 				C93A0FF42292232E0006C88F /* reset.eep in Resources */,
+				9BE10718275F4CFE00C708D5 /* wb32-dfu-updater_cli in Resources */,
```
Comment (fauxpark) [inline macos/QMK Toolbox.xcodeproj/project.pbxproj:377]: Move below `teensy_loader_cli`

File: macos/QMK Toolbox.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -267,6 +273,8 @@
 		3AC7BE50272565B600FED7DB /* Bootloader */ = {
 			isa = PBXGroup;
 			children = (
+				9B7999EE280AA1D0007469DA /* WB32DFUDevice.m */,
+				9B7999ED280AA1C8007469DA /* WB32DFUDevice.h */,
```
Comment (fauxpark) [inline macos/QMK Toolbox.xcodeproj/project.pbxproj:277]: Move below `USBTinyISPDevice`

File: macos/QMK Toolbox.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -221,6 +226,7 @@
 		3A53FB5B268A33F10020BAB0 /* Resources */ = {
 			isa = PBXGroup;
 			children = (
+				9BE10717275F4CFE00C708D5 /* wb32-dfu-updater_cli */,
```
Comment (fauxpark) [inline macos/QMK Toolbox.xcodeproj/project.pbxproj:229]: Move below `teensy_loader_cli`

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by wb-Joy** [comment]
The format has been adjusted.

**Comment by fauxpark** [comment]
Waiting for https://github.com/Homebrew/homebrew-core/pull/98739

File: macos/QMK Toolbox/AppDelegate.m [hunk 9]
```diff
@@ -62,6 +62,8 @@ - (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
     [self.logTextView logInfo:@" - Caterina (Arduino, Pro Micro) via avrdude (http://nongnu.org/avrdude/)"];
     [self.logTextView logInfo:@" - HalfKay (Teensy, Ergodox EZ) via Teensy Loader (https://pjrc.com/teensy/loader_cli.html)"];
     [self.logTextView logInfo:@" - LUFA/QMK HID via hid_bootloader_cli (https://github.com/abcminiuser/lufa)"];
+    [self.logTextView logInfo:@" - WB32 DFU via wb32-dfu-updater_cli (https://github.com/WestberryTech/wb32-dfu-updater)"];
+
```
Comment (fauxpark) [inline macos/QMK Toolbox/AppDelegate.m:66]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: macos/QMK Toolbox.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -148,6 +150,9 @@
 		3AE6DA022726CB7A00CECA21 /* USBAspDevice.m */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.objc; path = USBAspDevice.m; sourceTree = "<group>"; };
 		3AE6DA042726CBB200CECA21 /* USBTinyISPDevice.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = USBTinyISPDevice.h; sourceTree = "<group>"; };
 		3AE6DA052726CBB200CECA21 /* USBTinyISPDevice.m */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.objc; path = USBTinyISPDevice.m; sourceTree = "<group>"; };
+		9B7999ED280AA1C8007469DA /* WB32DFUDevice.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = WB32DFUDevice.h; sourceTree = "<group>"; };
+		9B7999EE280AA1D0007469DA /* WB32DFUDevice.m */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.objc; path = WB32DFUDevice.m; sourceTree = "<group>"; };
```
Comment (fauxpark) [inline macos/QMK Toolbox.xcodeproj/project.pbxproj:154]:
```suggestion
		9B7999ED280AA1C8007469DA /* WB32DFUDevice.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = WB32DFUDevice.h; sourceTree = "<group>"; };
		9B7999EE280AA1D0007469DA /* WB32DFUDevice.m */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.objc; path = WB32DFUDevice.m; sourceTree = "<group>"; };
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by wb-Joy** [comment]
> Waiting for [Homebrew/homebrew-core#98739](https://github.com/Homebrew/homebrew-core/pull/98739)

Has been merged.

**Comment by zvecr** [comment]
You still need to pull down the "official" binary from homebrew and update the PR.

**Comment by wb-Joy** [comment]
> You still need to pull down the "official" binary from homebrew and update the PR.

I've already done that.

**Comment by zvecr** [comment]
Ignore me, for whatever reason the sha256 didnt match. Just retried by steps and they do indeed match.

**Comment by wb-Joy** [comment]
> ~Ignore me, for whatever reason the sha256 didnt match. Just retried by steps and they do indeed match.~
> 
> You have uploaded the `monterey` bottle artifact, rather than the `catalina` ones.
> 
> executable should be `5becb921ab4cb3b1cce87418720fb3e8b5cc650eea0f0934815b2d291af2950a` but is `eca41b6ce4e88ff2c25eb8932adde0823a8e98c43426a3e8a859e3aa073da7b1`.

It's been re-uploaded.

**Comment by fauxpark** [comment]
Still wrong - the Homebrew placeholder needs to be changed.
This PR:
```
fauxpark@marissa:downloads$ otool -L wb32-dfu-updater_cli  
wb32-dfu-updater_cli:
	@@HOMEBREW_PREFIX@@/opt/libusb/lib/libusb-1.0.0.dylib (compatibility version 4.0.0, current version 4.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
fauxpark@marissa:downloads$ shasum -a256 wb32-dfu-updater_cli
5becb921ab4cb3b1cce87418720fb3e8b5cc650eea0f0934815b2d291af2950a  wb32-dfu-updater_cli
```

What it should be:
```
fauxpark@marissa:bin$ otool -L wb32-dfu-updater_cli  
wb32-dfu-updater_cli:
	@executable_path/../Frameworks/libusb-1.0.0.dylib (compatibility version 4.0.0, current version 4.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
fauxpark@marissa:bin$ shasum -a256 wb32-dfu-updater_cli
7bfd47724864c8042fb169aa6b6dd5ceef03ca477d29f9badd345d2503d7064a  wb32-dfu-updater_cli
```

See #342

**Comment by wb-Joy** [comment]
Okay, the placeholder have been changed. That should be all right.

File: macos/QMK Toolbox.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -52,6 +52,8 @@
 		3AE6DA002726CB2600CECA21 /* STM32DuinoDevice.m in Sources */ = {isa = PBXBuildFile; fileRef = 3AE6D9FF2726CB2600CECA21 /* STM32DuinoDevice.m */; };
 		3AE6DA032726CB7A00CECA21 /* USBAspDevice.m in Sources */ = {isa = PBXBuildFile; fileRef = 3AE6DA022726CB7A00CECA21 /* USBAspDevice.m */; };
 		3AE6DA062726CBB200CECA21 /* USBTinyISPDevice.m in Sources */ = {isa = PBXBuildFile; fileRef = 3AE6DA052726CBB200CECA21 /* USBTinyISPDevice.m */; };
+		9B7999EF280AA1D0007469DA /* WB32DFUDevice.m in Sources */ = {isa = PBXBuildFile; fileRef = 9B7999EE280AA1D0007469DA /* WB32DFUDevice.m */; };
+		9BE10718275F4CFE00C708D5 /* wb32-dfu-updater_cli in Resources */ = {isa = PBXBuildFile; fileRef = 9BE10717275F4CFE00C708D5 /* wb32-dfu-updater_cli */; };
 		3AFD4BD1281AB86800ADCB65 /* libhidapi.0.dylib in CopyFiles */ = {isa = PBXBuildFile; fileRef = 3AFD4BCF281AB83C00ADCB65 /* libhidapi.0.dylib */; settings = {ATTRIBUTES = (CodeSignOnCopy, ); }; };
```
Comment (fauxpark) [inline macos/QMK Toolbox.xcodeproj/project.pbxproj:57]:
```suggestion
		3AFD4BD1281AB86800ADCB65 /* libhidapi.0.dylib in CopyFiles */ = {isa = PBXBuildFile; fileRef = 3AFD4BCF281AB83C00ADCB65 /* libhidapi.0.dylib */; settings = {ATTRIBUTES = (CodeSignOnCopy, ); }; };
		9B7999EF280AA1D0007469DA /* WB32DFUDevice.m in Sources */ = {isa = PBXBuildFile; fileRef = 9B7999EE280AA1D0007469DA /* WB32DFUDevice.m */; };
		9BE10718275F4CFE00C708D5 /* wb32-dfu-updater_cli in Resources */ = {isa = PBXBuildFile; fileRef = 9BE10717275F4CFE00C708D5 /* wb32-dfu-updater_cli */; };
```

File: macos/QMK Toolbox.xcodeproj/project.pbxproj [hunk 12]
```diff
@@ -148,6 +150,9 @@
...
 		3AE6DA042726CBB200CECA21 /* USBTinyISPDevice.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = USBTinyISPDevice.h; sourceTree = "<group>"; };
 		3AE6DA052726CBB200CECA21 /* USBTinyISPDevice.m */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.objc; path = USBTinyISPDevice.m; sourceTree = "<group>"; };
+		9B7999ED280AA1C8007469DA /* WB32DFUDevice.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = WB32DFUDevice.h; sourceTree = "<group>"; };
+		9B7999EE280AA1D0007469DA /* WB32DFUDevice.m */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.objc; path = WB32DFUDevice.m; sourceTree = "<group>"; };
+		9BE10717275F4CFE00C708D5 /* wb32-dfu-updater_cli */ = {isa = PBXFileReference; lastKnownFileType = "compiled.mach-o.executable"; path = "wb32-dfu-updater_cli"; sourceTree = "<group>"; };
 		3AFD4BCF281AB83C00ADCB65 /* libhidapi.0.dylib */ = {isa = PBXFileReference; lastKnownFileType = "compiled.mach-o.dylib"; path = libhidapi.0.dylib; sourceTree = "<group>"; };
```
Comment (fauxpark) [inline macos/QMK Toolbox.xcodeproj/project.pbxproj:156]:
```suggestion
		3AFD4BCF281AB83C00ADCB65 /* libhidapi.0.dylib */ = {isa = PBXFileReference; lastKnownFileType = "compiled.mach-o.dylib"; path = libhidapi.0.dylib; sourceTree = "<group>"; };
		9B7999ED280AA1C8007469DA /* WB32DFUDevice.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = WB32DFUDevice.h; sourceTree = "<group>"; };
		9B7999EE280AA1D0007469DA /* WB32DFUDevice.m */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.objc; path = WB32DFUDevice.m; sourceTree = "<group>"; };
		9BE10717275F4CFE00C708D5 /* wb32-dfu-updater_cli */ = {isa = PBXFileReference; lastKnownFileType = "compiled.mach-o.executable"; path = "wb32-dfu-updater_cli"; sourceTree = "<group>"; };
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
One last thing.

**Review by fauxpark — APPROVED** [review]

**Pull request merged**